### DISCUSSION TITLE
Can't figure out how to stream markdown into tiptap

### DISCUSSION THREAD

**Discussion by fgatti675** [ueberdosis/tiptap#5563] [General Questions] [answered]
Hi everyone,
I am having a hard time figuring out how to stream markdown generated from an LLM into tiptap.
If I insert content as it comes, it fails in many ways. If I wait til the result is complete and just insert the full string, it works perfectly.
I have been trying a middle ground, spliting the markdown by lines, so I make sure format like **bold** are handled correctly.

This code simulates the calls that my code makes while streaming:
```
editor.chain().focus()
    .insertContent("- **Autosave**: Never worry about losing your work… are saved as you type, giving you peace of mind.\n")
    .run()
editor.chain().focus()
    .insertContent("- **Another line**: Never worry about losing your work… are saved as you type, giving you peace of mind.\n")
    .run();
```
And how it looks like:

<img width="607" alt="Screenshot 2024-08-28 at 01 29 55" src="https://github.com/user-attachments/assets/8d243e3a-b831-43b8-a3c4-97c2d9900759">

I am using the markdown extension configured like:
```
export const markdownExtension = Markdown.configure({
    html: true,
});
```
It generates the new bulletlist node under the last one, no matter what I do, calling `.enter()` does nothing apparently, though I was expecting to jump to a new line

Any help is greatly appreciated!

**Comment by nperez0111** [comment]
It is very difficult to get streaming of formatted content into the editor it took me two weeks to get everything right for the content Ai pro extension. I would not say that there is an easy answer to this at all. I would recommend streaming into a preview element instead and only inserting the content into the editor afterwards.

**Reply by fgatti675** [reply]
Thank you. Yes I realised it is pretty complicated.
I will try your approach, I am going to stream HTML instead of markdown, directly into a decorator, and replace with the final content when complete. Does that make sense?
Thanks for your help!

**Reply by nperez0111** [reply]
Yea into a decoration sounds like a good plan

**Reply by fgatti675** [reply]
I could manage with your help, thanks a lot!!

https://github.com/user-attachments/assets/29f8e7e9-1a28-401d-ae77-d6f499d2a95d

**Reply by anasmir-cm** [reply]
hey @fgatti675 can you share your solution for this?

**Comment by whisper-bye** [comment]
@fgatti675 would you please share your solution? thx a lot!

**Reply by gjc14** [reply]
Hi maybe this will be helpful. I made a component based on https://tiptap.dev/docs/editor/extensions/custom-extensions/node-views/react to handle api call and stream render.

[gjc14/tiptap-streaming-preview](https://github.com/gjc14/tiptap-streaming-preview)

**Reply by whisper-bye** [reply]
@gjc14 amazing!!!

**Comment by eRuaro** [comment]
if the LLM is returning markdown content, you can use a markdown to html converter and pass the results to `setContent`:
```
import React, { useState, useEffect } from 'react';
import { useEditor, EditorContent } from '@tiptap/react';
import StarterKit from '@tiptap/starter-kit';
import Markdown from '@tiptap/extension-markdown';
import Link from '@tiptap/extension-link';
import { marked } from 'marked';

const MarkdownEditor = () => {
 const [answer, setAnswer] = useState([]);
 const [streaming, setStreaming] = useState(false);

  ... code that handles llm streaming, presumably this appends to the answer array

 const markdownToHtml = (markdown) => {
   return marked(markdown, {
     breaks: true,    // Convert \n to <br>
     gfm: true,       // Enable GitHub Flavored Markdown
     async: false
   });
 };

 const answerEditor = useEditor({
   editable: false,
   extensions: [
     StarterKit,
     Markdown,
     Link.configure({
       autolink: false,
     })
   ],
 });

 useEffect(() => {
   const content = answer.join("");
   const htmlContent = markdownToHtml(content);
   answerEditor?.commands.setContent(htmlContent);
 }, [answer, answerEditor]);

 return (
   <>
     {streaming && (
       <div className="border border-gray-600 shadow-md rounded-lg px-4 py-1 flex flex-col">
         <EditorContent 
           editor={answerEditor} 
           className="prose prose-p:text-md max-w-none whitespace-pre-wrap flex-grow" 
         />
       </div>
     )}
   </>
 );
};

export default MarkdownEditor; 
```

**Comment by luckymore** [comment]
Alas, it must be said that built-in features like platejs are actually charged in tiptap...

**Reply by bdbch** [reply]
What exact feature do you mean? The conversion service (server-side) or the client-side implementation Tiptap now supports via it's `@tiptap/markdown` package?

**Reply by luckymore** [reply]
Such as enhanced tables and AI assistance.

**Comment by bdbch** [comment, marked as answer]
Coming in with a new answer for everyone who looks for Markdown Support in Tiptap. Markdown is officially supported via an extra **open source** package `@tiptap/markdown`. Read more [here](https://tiptap.dev/docs/editor/markdown).

**Reply by oskarscholander** [reply]
I use the Markdown extension as of today and currently am building a LLM chat. The LLM returns markdown, but inserting it into my custom suggestion extension (based on decorations) doesn't make the markdown formatted. It is not even formatted when I apply the suggestion into the actual Editor content.

Do I need a separate markdown->html converter, or should I only use tiptap/markdown, if so how?

**Reply by bdbch** [reply]
How are you inserting content into the editor? If you use `editor.commands.insertContent()` or `editor.commands.insertContentAt()` with the `contentType: 'markdown'` option it should be able to parse. The only thing that is important while streaming is, that you replace the whole range of streamed content so you don't only let single tokens parse in.

For example you stream in the content `Hello **World**` and it's streamed in the chunks `Hel`, `lo`, `**Wo`. `rld**` - you should not insert each of these slices alone but replace the already inserted content so for this example. The markdown extension can only parse on the insert command, it's not constantly scanning the content whether there is valid markdown or not.

**Reply by bdbch** [reply]
Here is an example:
https://codesandbox.io/p/sandbox/vhzcq3

It's not perfect because you'd need a more stable way to find the end-position of the insertion after the content was inserted. In the future I plan to have a simpler solution to stream content into the editor.

**Reply by oskarscholander** [reply]
I might very well do this totally wrong. My Suggestion extension adds Decorations into the Editor like this. Very simplified. I do not like this approach at all. The innerHTML is very wacky. I guess I need to rewrite so I use editor.commands instead? Thanks for your example!

```
const container = document.createElement("span");
container.innerHTML = suggestion.suggestedContent;

decorations.push(
    Decoration.widget(suggestion.from, container, {
        side: -1, // Position before the hidden text (left side)
    })
);
```

Then I use state.apply to add them to the Editor.

**Reply by bdbch** [reply]
Depends on what you want. My example is straight inserting the content into the editor. If you want to have some kind of suggestion extension, that shows a preview first & accepts via tab for example, decorations are the right way to do this.

However you have to manage the decorations yourself as you can't just pass in ProseMirror content there.

You can use the `MarkdownManager` that is exported from the markdown package to parse incoming markdown to Prosemirror JSON and then use this to create decoration HTML (one of my ideas but not 100% sure).

Another solution would be to just insert the content but mark it as a suggestion via marks / node attributes - but handling suggestions is very tricky.

**Reply by oskarscholander** [reply]
Tricky, but quite fun actually. The only concern is dangerous HTML injections. Based on what u share I guess I need to use innerHTML, but make sure to sanitize it?

**And yes, my suggestion supports:**
* Replace original text with suggested text
* Custom background color and hover effect
* Custom sticky menu for Accept and Reject

I will immediately try out ´MarkdownManager´ and report back. Thanks!

**Reply by bdbch** [reply]
Yep you should sanitize what comes back from the LLM. Then you can let the content be parsed by the MarkdownManager and get Tiptap JSON from it - which you can then use to render suggestions on your doc.

As said I'm not sure what your exact use case is.

**Reply by oskarscholander** [reply]
I did a repro now. Actually now when I see it standalone like this it really feels like I am allllmost there:
https://codesandbox.io/p/sandbox/tiptap-suggestions-pn9l7d

Kinda cool right? Since I do not know how to make it work with editor.commands I use https://www.npmjs.com/package/marked to convert markdown to HTML. I have yet to sanitize it. Only issue I have is that it doesn't convert to HTML when I apply. Do you have any suggestions to this?

**Reply by bdbch** [reply]
You would need to run `insertContentAt()` with `{ contentType: 'markdown' }` there.

**Reply by bdbch** [reply]
Here's a fork:
https://codesandbox.io/p/sandbox/tiptap-suggestions-forked-qv8k97

**Reply by oskarscholander** [reply]
omg thank you so much!! I have been working on this all day even with a fever and had this thing that I was worried about that it might not will be working at all, and here you come and put the last cheddar on it! This made my day! Thanks buddy!

I didn't realize that I have access to `chain`! Of course!

**Reply by bdbch** [reply]
No worries, good luck with your project!

**Reply by yazan-equ** [reply]
@bdbch you should add streaming LLM to documentation, you will a lot more users!
